chalice/chalicelib/admin.py

```diff
@@ -229,3 +230,99 @@ def manual_update_teams(event, context):
+    that adds significant processing load for little benefit.
+    By regenerating the stats on a schedule we can make the interface much faster and keep the database
+    processing load lighter.
+    """
+    start_time = time()
+    elapsed_time = 0
+    status = 0
+    deleted_audits = 0
+    remove_count = 100
+    # keep 1 year
+    time_limit_days = 365
+    # run for 4 minutes
+    # (lambda timeout is set to 5)
+    execution_limit = 240
+    commands = []
+    try:
+        dbh = DatabaseHandle(app)
+        db = dbh.get_handle()
+        app.log.debug(f"Listing oldest records")
+
+        select_expired_audit_ids = f"""
+            SELECT id
+            FROM public.account_audit AS all_audits
+            WHERE DATE_PART('day', CURRENT_TIMESTAMP - all_audits.date_completed) > {time_limit_days}
+            ORDER BY id
+            LIMIT {remove_count} OFFSET 0;
+            """
+        audit_cursor = db.execute_sql(select_expired_audit_ids)
+        delete_statements = []
+        for audit_row in audit_cursor.fetchall():
+            account_audit_id = audit_row[0]
+            select_expired_resource_ids = f"""
+                SELECT id
+                FROM public.audit_resource
+                WHERE account_audit_id = {account_audit_id};
+                """
+            resource_cursor = db.execute_sql(select_expired_resource_ids)
+            app.log.debug(f"Deleting audit: {account_audit_id}")
+            for resource_row in resource_cursor.fetchall():
+                resource_id = resource_row[0]
+                delete_compliance = f"""
+                    DELETE
+                    FROM public.resource_compliance
+                    WHERE audit_resource_id = {resource_id};
+                    """
+                delete_statements.append(delete_compliance)
+
+            delete_resource = f"""
+                DELETE
+                FROM public.audit_resource
+                WHERE account_audit_id = {account_audit_id};
+                """
+            delete_statements.append(delete_resource)
+
+            delete_criterion = f"""
+                DELETE
+                FROM public.audit_criterion
+                WHERE account_audit_id = {account_audit_id};
+                """
+            delete_statements.append(delete_criterion)
+
+            delete_audit = f"""
+                DELETE
+                FROM public.account_audit
+                WHERE id = {account_audit_id};
+                """
+            delete_statements.append(delete_audit)
+            app.log.debug(f"Running {len(delete_statements)} deletes");
+            dbh.execute_commands(delete_statements, "csw")
+            deleted_audits += 1
+            app.log.debug(f"Deleted {deleted_audits} audits so far");
+            elapsed_time = time() - start_time;
+
+            if elapsed_time > execution_limit:
+                break;
```

Comment:
should we do something special if we're at risk of the lambda timing out and the expired audits haven't all been deleted? (write an error somewhere / retrigger the lambda) rather than just breaking the loop and exiting - won't that mean we could build up a backlog of expired audits or is that not likely

Reply:
Yeah - It definitely can't do them all in one go because of the execution time limit on the lambda. 

I'd forgotten to change the frequency. I've just pushed a commit to run this every 15 minutes so it should slowly catch up with itself and then it should be able to keep on top of the deletions each day. The earliest ones there should only be a few audits each day because there weren't many subscribed accounts. I guess we might have to run it more frequently if we keep growing the number of accounts. This should handle <190 accounts maybe.

Hopefully we'll replace CSW with the Config Rules service before that becomes a problem.

Reply:
I guess I could do something else to notify us if the oldest audit is >365 days ago which should tell us if it's not keeping up.

Reply:
I'll have a think about how to do that and raise it as a separate issue maybe

Reply:
☑️ https://github.com/alphagov/csw-backend/issues/183